In [47]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np

# Part 1: Load the dataset

In [48]:
# Load the given datasets
categorical_data = pd.read_csv('chronic_kidney_disease_categorical.csv')
numerical_data = pd.read_csv('chronic_kidney_disease_numerical.csv')
# Print the data
print("Categorical Dataset:")
print(categorical_data)

print("\nNumerical Dataset:")
print(numerical_data)

Categorical Dataset:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no 

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:The Chronic Kidney Disease(CKD) dataset contains 400 instances and 24 features. It can be used in predicting the CKD. The features are in both categorical and numerical. The features are age, blood pressure, hemoglobin, suger and so on. The labels are ckd and notckd. Categorical Features: specific gravity, albumin, Pus Cell clumps. Numerical features: age, blood pressure, Blood Glucose Random.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [49]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicate_numerical_rows = numerical_data[numerical_data.duplicated()]
print(f"Number of duplicate rows in numerical dataset: {duplicate_numerical_rows.shape[0]}")

# Drop these duplicate rows
numerical_data_cleaned = numerical_data.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicate_categorical_rows = categorical_data[categorical_data.duplicated()]
print(f"Number of duplicate rows in categorical dataset: {duplicate_categorical_rows.shape[0]}")
categorical_data_cleaned = categorical_data.drop_duplicates()


Number of duplicate rows in numerical dataset: 30
Number of duplicate rows in categorical dataset: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No. Not all the unique ids are present in both datasets. This is because there may exist missing data. The rows with missing ids will have NaN values. 

In [50]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined = pd.merge(numerical_data, categorical_data, on='unique_id', how='outer')
#Print the combined dataset
print(combined)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
487     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
488     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
489     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
490     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
491     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [51]:
# Calculate the percentage of rows that contain atleast one missing value
total_rows = combined.shape[0]
num_contain_missing_value = combined.isnull().any(axis = 1).sum()
percentage = (num_contain_missing_value / total_rows) * 100

# Print %
print(f"{percentage:.2f}%")
# Drop these rows from the dataset
dropped_data = combined.dropna()
# Print the Dataset
print(dropped_data)

59.76%
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
482     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
483     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
486     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
487     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
488     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn  

## Part 3.4: Sort the dataset according to the Labels

In [52]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_data = combined.sort_values(by='Target')
sorted_data.reset_index(drop=True, inplace=True)
# Print the dataset
print(sorted_data)

     unique_id   age    bp    bgr     bu    sc    sod   pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0   7.6   7.1  22.0  ...   
1       619200  68.0  80.0  171.0   30.0   1.0    NaN   NaN  13.7  43.0  ...   
2       617089  67.0  80.0  182.0  391.0  32.0  163.0  39.0   NaN   NaN  ...   
3       615697   6.0  60.0   94.0   67.0   1.0  135.0   4.9   9.9  30.0  ...   
4       608902  56.0  90.0  176.0  309.0  13.3  124.0   6.5   3.1   9.0  ...   
..         ...   ...   ...    ...    ...   ...    ...   ...   ...   ...  ...   
487     731540  63.0  70.0  130.0   37.0   0.9  150.0   5.0  13.4  41.0  ...   
488     261137  35.0  60.0  105.0   39.0   0.5  135.0   3.9  14.7  43.0  ...   
489     379507  17.0  60.0  114.0   50.0   1.0  135.0   4.9  14.2  51.0  ...   
490     721270  58.0  70.0  102.0   48.0   1.2  139.0   4.3  15.0  40.0  ...   
491     441170  79.0  80.0  111.0   44.0   1.2  146.0   3.6  16.3  40.0  ...   

           pc         pcc          ba  

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [53]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_data = pd.get_dummies(sorted_data, columns = categorical_columns)
# Print the dataset
print(encoded_data)

     unique_id   age    bp    bgr     bu    sc    sod   pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0   7.6   7.1  22.0  ...   
1       619200  68.0  80.0  171.0   30.0   1.0    NaN   NaN  13.7  43.0  ...   
2       617089  67.0  80.0  182.0  391.0  32.0  163.0  39.0   NaN   NaN  ...   
3       615697   6.0  60.0   94.0   67.0   1.0  135.0   4.9   9.9  30.0  ...   
4       608902  56.0  90.0  176.0  309.0  13.3  124.0   6.5   3.1   9.0  ...   
..         ...   ...   ...    ...    ...   ...    ...   ...   ...   ...  ...   
487     731540  63.0  70.0  130.0   37.0   0.9  150.0   5.0  13.4  41.0  ...   
488     261137  35.0  60.0  105.0   39.0   0.5  135.0   3.9  14.7  43.0  ...   
489     379507  17.0  60.0  114.0   50.0   1.0  135.0   4.9  14.2  51.0  ...   
490     721270  58.0  70.0  102.0   48.0   1.2  139.0   4.3  15.0  40.0  ...   
491     441170  79.0  80.0  111.0   44.0   1.2  146.0   3.6  16.3  40.0  ...   

     cad_yes  appet_good  appet_poor  p

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [54]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for column in numerical_columns:
    mean = encoded_data[column].mean()
    std_devi = encoded_data[column].std()
    outlier_upper = mean + 3 * std_devi
    outlier_lower = mean - 3 * std_devi
    encoded_data = encoded_data[(encoded_data[column] >= outlier_lower) & (encoded_data[column] <=outlier_upper)]

# Print the dataset
print(encoded_data)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
16      580048  61.0  80.0  360.0   19.0  0.70  137.0  4.4  15.2  44.0  ...   
20      621332  59.0  80.0  303.0   35.0  1.30  122.0  3.5  10.4  35.0  ...   
27      679864  53.0  60.0   91.0  114.0  3.25  142.0  4.3   8.6  28.0  ...   
38      640519  54.0  70.0  111.0  146.0  7.50  141.0  4.7  11.0  35.0  ...   
40      631648  46.0  80.0  160.0   40.0  2.00  140.0  4.1   9.0  27.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
487     731540  63.0  70.0  130.0   37.0  0.90  150.0  5.0  13.4  41.0  ...   
488     261137  35.0  60.0  105.0   39.0  0.50  135.0  3.9  14.7  43.0  ...   
489     379507  17.0  60.0  114.0   50.0  1.00  135.0  4.9  14.2  51.0  ...   
490     721270  58.0  70.0  102.0   48.0  1.20  139.0  4.3  15.0  40.0  ...   
491     441170  79.0  80.0  111.0   44.0  1.20  146.0  3.6  16.3  40.0  ...   

     cad_yes  appet_good  appet_poor  pe_no  pe_yes

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [55]:
# Normalize the all Numerical Attributes in the dataset.
#numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for column in numerical_columns:
    min_val = encoded_data[column].min()
    max_val = encoded_data[column].max()
    # Apply the Min-Max normalization formula
    encoded_data[column] = (encoded_data[column] - min_val) / (max_val - min_val)
# Print the dataset
print(encoded_data)

     unique_id       age    bp       bgr        bu        sc       sod  \
16      580048  0.628205  0.50  0.935484  0.058824  0.026087  0.566667   
20      621332  0.602564  0.50  0.751613  0.163399  0.078261  0.066667   
27      679864  0.525641  0.00  0.067742  0.679739  0.247826  0.733333   
38      640519  0.538462  0.25  0.132258  0.888889  0.617391  0.700000   
40      631648  0.435897  0.50  0.290323  0.196078  0.139130  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
487     731540  0.653846  0.25  0.193548  0.176471  0.043478  1.000000   
488     261137  0.294872  0.00  0.112903  0.189542  0.008696  0.500000   
489     379507  0.064103  0.00  0.141935  0.261438  0.052174  0.500000   
490     721270  0.589744  0.25  0.103226  0.248366  0.069565  0.633333   
491     441170  0.858974  0.50  0.132258  0.222222  0.069565  0.866667   

          pot      hemo       pcv  ...  cad_yes  appet_good  appet_poor  \
16   0.500000  0.773913  0.714286  .

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:The unique_ID column could be removed from the dataset. The unique_ID column is to identify each observation. Thus, it does not have any meaningful relationship with the target variable. 

In [56]:
#Remove that column
encoded_data = encoded_data.drop(columns=['unique_id'])
# Print the dataset
print(encoded_data)

          age    bp       bgr        bu        sc       sod       pot  \
16   0.628205  0.50  0.935484  0.058824  0.026087  0.566667  0.500000   
20   0.602564  0.50  0.751613  0.163399  0.078261  0.066667  0.263158   
27   0.525641  0.00  0.067742  0.679739  0.247826  0.733333  0.473684   
38   0.538462  0.25  0.132258  0.888889  0.617391  0.700000  0.578947   
40   0.435897  0.50  0.290323  0.196078  0.139130  0.666667  0.421053   
..        ...   ...       ...       ...       ...       ...       ...   
487  0.653846  0.25  0.193548  0.176471  0.043478  1.000000  0.657895   
488  0.294872  0.00  0.112903  0.189542  0.008696  0.500000  0.368421   
489  0.064103  0.00  0.141935  0.261438  0.052174  0.500000  0.631579   
490  0.589744  0.25  0.103226  0.248366  0.069565  0.633333  0.473684   
491  0.858974  0.50  0.132258  0.222222  0.069565  0.866667  0.289474   

         hemo       pcv      wbcc  ...  cad_yes  appet_good  appet_poor  \
16   0.773913  0.714286  0.394737  ...    False 

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [57]:
# Export the dataframe to a new csv file
encoded_data.to_csv('cleaned_dataset.csv', index=False)
# Export the dataframe to a new json file
encoded_data.to_json('cleaned_dataset.json', orient='records', lines=True)
#print(encoded_data.head(15))

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here
| Age |   BP     |   BGR    |    BU    |    SC    |   SOD    |   POT    |   Hemo   |   PCV    |   WBCC   | ...   | Cad Yes | Appet Good | Appet Poor |  PE No  | PE Yes | Ane No | Ane Yes | Target CKD | Target CKD\t | Target Not CKD |
|-----|----------|----------|----------|----------|----------|----------|----------|----------|----------|-------|---------|------------|------------|---------|--------|--------|---------|-------------|--------------|----------------|
|  16 | 0.628205 | 0.50     | 0.935484 | 0.058824 | 0.026087 | 0.566667 | 0.773913 | 0.714286 | 0.394737 | ...   | False   | True       | False      | True    | False  | True   | False   | True        | False        | False          |
|  20 | 0.602564 | 0.50     | 0.751613 | 0.163399 | 0.078261 | 0.066667 | 0.356522 | 0.457143 | 0.622807 | ...   | False   | False      | True       | True    | False  | True   | False   | True        | False        | False          |
|  27 | 0.525641 | 0.00     | 0.067742 | 0.679739 | 0.247826 | 0.733333 | 0.200000 | 0.257143 | 0.631579 | ...   | False   | False      | True       | False   | True   | False  | True    | True        | False        | False          |
|  38 | 0.538462 | 0.25     | 0.132258 | 0.888889 | 0.617391 | 0.700000 | 0.408696 | 0.457143 | 0.421053 | ...   | False   | True       | False      | True    | False  | True   | False   | True        | False        | False          |
|  40 | 0.435897 | 0.50     | 0.290323 | 0.196078 | 0.139130 | 0.666667 | 0.234783 | 0.228571 | 0.377193 | ...   | False   | False      | True       | True    | False  | False  | True    | True        | False        | False          |
|  45 | 0.628205 | 0.75     | 0.287097 | 0.189542 | 0.095652 | 0.433333 | 0.434783 | 0.428571 | 0.508772 | ...   | False   | False      | True       | True    | False  | True   | False   | True        | False        | False          |
|  47 | 0.474359 | 1.00     | 0.190323 | 0.967320 | 0.991304 | 0.066667 | 0.156522 | 0.142857 | 0.508772 | ...   | False   | False      | True       | False   | True   | False  | True    | True        | False        | False          |
|  49 | 0.551282 | 1.00     | 0.732258 | 0.281046 | 0.208696 | 0.633333 | 0.426087 | 0.428571 | 0.859649 | ...   | False   | True       | False      | True    | False  | True   | False   | True        | False        | False          |
|  54 | 0.358974 | 0.50     | 0.019355 | 0.993464 | 0.800000 | 0.700000 | 0.113043 | 0.142857 | 0.000000 | ...   | False   | True       | False      | True    | False  | False  | True    | True        | False        | False          |
|  57 | 0.884615 | 0.00     | 0.251613 | 0.189542 | 0.147826 | 0.900000 | 0.400000 | 0.457143 | 0.491228 | ...   | True    | False      | True       | False   | True   | True   | False   | True        | False        | False          |
|  59 | 0.705128 | 0.25     | 0.103226 | 0.248366 | 0.243478 | 0.566667 | 0.486957 | 0.428571 | 0.289474 | ...   | False   | True       | False      | False   | True   | True   | False   | True        | False        | False          |
|  61 | 0.448718 | 1.00     | 0.432258 | 0.124183 | 0.052174 | 0.633333 | 0.295652 | 0.400000 | 0.473684 | ...   | False   | True       | False      | True    | False  | False  | True    | True        | False        | False          |
|  62 | 0.666667 | 1.00     | 0.300000 | 0.287582 | 0.591304 | 0.666667 | 0.139130 | 0.200000 | 0.324561 | ...   | False   | True       | False      | False   | True   | True   | False   | True        | False        | False          |
|  63 | 0.115385 | 0.75     | 0.119355 | 0.196078 | 0.113043 | 0.166667 | 0.173913 | 0.114286 | 0.754386 | ...   | False   | True       | False      | True    | False  | False  | True    | True        | False        | False          |
|  68 | 0.615385 | 0.00     | 0.703226 | 0.169935 | 0.113043 | 0.333333 | 0.139130 | 0.171429 | 1.000000 | ...   | False   | False      | True       | True    | False  | False  | True    | True        | False        | False          |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?
We can use external tools to mitigate hallucinations. After the LLM conversions, put the output through dedicated data validation libraries. For example, we can use pandas when convert data to CSV or JSON.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [58]:
# Code to rename all the columns in the dataset
rename_dict = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class',
}

encoded_data.rename(columns=rename_dict, inplace=True)
print(encoded_data)

          age  blood pressure  blood glucose random  blood urea  \
16   0.628205            0.50              0.935484    0.058824   
20   0.602564            0.50              0.751613    0.163399   
27   0.525641            0.00              0.067742    0.679739   
38   0.538462            0.25              0.132258    0.888889   
40   0.435897            0.50              0.290323    0.196078   
..        ...             ...                   ...         ...   
487  0.653846            0.25              0.193548    0.176471   
488  0.294872            0.00              0.112903    0.189542   
489  0.064103            0.00              0.141935    0.261438   
490  0.589744            0.25              0.103226    0.248366   
491  0.858974            0.50              0.132258    0.222222   

     serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
16           0.026087  0.566667   0.500000    0.773913            0.714286   
20           0.078261  0.066667   0.263

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [62]:
# Converted SQL to Pandas code
result = encoded_data.groupby('age').size().reset_index(name='count')
print(result)

         age  count
0   0.000000      1
1   0.038462      2
2   0.064103      1
3   0.102564      1
4   0.115385      1
5   0.128205      1
6   0.141026      4
7   0.153846      1
8   0.166667      2
9   0.205128      1
10  0.217949      2
11  0.230769      8
12  0.256410      2
13  0.269231      5
14  0.282051      3
15  0.294872      6
16  0.307692      1
17  0.320513      2
18  0.333333      5
19  0.346154      2
20  0.358974      3
21  0.371795      3
22  0.384615      6
23  0.397436      6
24  0.410256      4
25  0.423077      5
26  0.435897      5
27  0.448718      6
28  0.461538      7
29  0.474359      2
30  0.487179      2
31  0.500000      3
32  0.512821      6
33  0.525641      2
34  0.538462      3
35  0.551282     13
36  0.564103      5
37  0.576923      8
38  0.589744      7
39  0.602564      3
40  0.615385     10
41  0.628205      5
42  0.641026      5
43  0.653846      6
44  0.666667      7
45  0.679487      7
46  0.692308      2
47  0.705128      1
48  0.717949      2


*Describe what the above code does here*
encoded_data.groupby('age'): groups the rows for encoded_data based on the values in 'age' column.
size(): this counts the number of rows in each group after encoded_data.groupby('age')
reset_index(name='count'): this convert the result into a DataFrame, and reset the index. Then it create a column called 'count', and it contains the count of rows for each value in the 'age'column.